### PR TITLE
executor: add recover mechanism for join workers

### PR DESCRIPTION
**Pull request by zz-jason**

### CODE REVIEW

**Comment by zz-jason** [comment]
@coocood @tiancaiamao PTAL

File: executor/join.go [hunk 1]
```diff
@@ -294,6 +294,10 @@ func (e *HashJoinExec) fetchOuterChunks(goCtx goctx.Context) {
 			close(e.outerResultChs[i])
 		}
 		e.workerWaitGroup.Done()
+
+		if r := recover(); r != nil {
```
Comment (tiancaiamao) [inline executor/join.go]: This should before `e.workerWaitGroup.Done()`.
When `WaitGroup.Done()` is called, the `e.joinResultCh` maybe close immediately, and send to a closed channel will panic.

File: executor/join.go [hunk 2]
```diff
@@ -560,7 +564,12 @@ func (e *HashJoinExec) runJoinWorker(workerID int) {
 }
 
 func (e *HashJoinExec) runJoinWorker4Chunk(workerID int) {
-	defer e.workerWaitGroup.Done()
+	defer func() {
+		e.workerWaitGroup.Done()
+		if r := recover(); r != nil {
```
Comment (tiancaiamao) [inline executor/join.go]: ditto

**Comment by coocood** [comment]
LGTM

**Comment by tiancaiamao** [comment]
LGTM
/run-all-tests

**Pull request merged**